include/utility/logger.h

```diff
@@ -0,0 +1,271 @@
+/***
+ * @file log.h
+ * @author zhuzhile08 (zhuzhile08@gmail.com)
+ *
+ * @brief header with loggers
+ * @brief function-esque macros for debugging
+ * @brief logs messages and exceptions with coloring
+ * @brief ANSI color codes
+ *
+ * @date 2022-02-03
+ *
+ * @copyright Copyright (c) 2022
+ *
+ *************************/
+
+#pragma once
+
+#include <iostream>
+#include <fstream>
+#include <utility>
+
+namespace utility {
+
+#define NODISCARD [[nodiscard]]
+
+class Logger {
+private:
+	/**
+	 * @brief basic ANSI fonts
+	 */
+	enum class Font : int {
+		NON = 0,
+		BLD = 1,
+		DIM = 2,
+		ITC = 3,
+		UNL = 4
+	};
+
+	/**
+	 * @brief cast a font enum value to an integer
+	 * 
+	 * @param font font to cast
+	 * 
+	 * @return constexpr int
+	 */
+	NODISCARD constexpr int font_cast(Font font) {
+		return static_cast<int>(font);
+	}
+
+	/**
+	 * @brief basic ANSI colors
+	 */
+	enum class Color : int {
+		BLK = 30,
+		RED = 31,
+		GRN = 32,
+		YEL = 33,
+		BLU = 34,
+		MAG = 35,
+		CYN = 36,
+		GRY = 90,
+		B_RED = 91,
+		B_GRN = 92,
+		B_YEL = 93,
+		B_BLU = 94,
+		B_MAG = 95,
+		B_CYN = 96,
+		WHT = 97,
+		DEF = WHT
+	};
+
+	/**
+	 * @brief cast a color enum value to an integer
+	 * 
+	 * @param color color to cast
+	 * 
+	 * @return constexpr int
+	 */
+	NODISCARD constexpr int color_cast(Color color) {
+		return static_cast<int>(color);
+	}
+
+	/**
+	 * @brief print an ANSI escape code to set font and color
+	 * 
+	 * @param font font
+	 * @param color color
+	 */
+	void ANSI(Font font, Color color) {
+		std::cout << "\033[" << font_cast(font) << ";" << color_cast(color) << "m";
+	}
+	/**
+	 * @brief set the color to default
+	 */
+	void set_color_default() {
+		ANSI(Font::NON, Color::WHT);
+	}
+
+public:
+	/**
+	 * @brief log normal messages
+	 *
+	 * @tparam ...Args types to print
+	 * @param ...message messages
+	*/
+	template <typename ... Args> void log(Args... message) {
+#ifndef NDEBUG
+		// print the message
+		(std::cout << ... << std::forward<Args>(message)) << end_l();
+#ifdef LYRA_LOG_FILE
+		(m_logFile << ... << std::forward<Args>(message)) << end_l();
+#endif
+#endif
+	}
+
+	/**
+	 * @brief log debug messages
+	 *
+	 * @tparam ...Args types to print
+	 * @param ...message messages
+	*/
+	template <typename ... Args> void debug(Args... message) {
+#ifndef NDEBUG
+		// set the color and font of the message
+		ANSI(Font::NON, Color::GRY);
+		// print the message
+		std::cout << "[DEBUG]: ";
+		(std::cout << ... << std::forward<Args>(message)) << end_l();
```

Comment:
This isn't thread safe

Reply:
Yes, I am aware of this major flaw, but I think this should do just fine


For now at least...


Jokes aside, I'll probably make a thread safe version of this the next few days, so yeah.